File: code/14-scripting/example_14.25/skill.py
#! /usr/bin/python
# Example 14.25 (Figure 14.6)

# This assumes that your /usr/bin/python is Python 3.
# To use, start a bunch of process you don't care about (loop.c in this
# directory makes a good candidate) then run: ./skill.py <pattern>

import os, re, subprocess, sys, time
if len(sys.argv) != 2:
    sys.stderr.write('usage: ' + sys.argv[0] + ' pattern\n')
    sys.exit(1)
ps = subprocess.Popen(['/bin/ps', '-w', '-w', '-x',
                       '-o', 'pid,command'], stdout=subprocess.PIPE)
toss = ps.stdout.readline()         # discard header line
for line in ps.stdout:
    line = line.decode().rstrip()
    proc = int(re.search('\S+', line).group())
    if re.search(sys.argv[1], line) and proc != os.getpid():
        print(line + '? ', flush=True, end='')
        answer = sys.stdin.readline()
        while not re.search('^[yn]', answer, re.I):
            print('? ', flush=True, end='')
            answer = sys.stdin.readline()
        if re.search('^y', answer, re.I):
            os.kill(proc, 9)
            time.sleep(1)
            try:                    # expect exception if process
                os.kill(proc, 0)    # no longer exists
                sys.stderr.write('unsuccessful; sorry\n');
                sys.exit(1)
            except: pass            # do nothing on expected exception

File: code/14-scripting/example_14.27/skill.rb
#! /usr/bin/ruby
# Example 14.27 (Figure 14.7)

# To use, start a bunch of process you don't care about (loop.c in this
# directory makes a good candidate) then run: ./skill.rb <pattern>

ARGV.length() == 1 or begin
    $stderr.print("usage: #{$0} pattern\n"); exit(1)
end

pat = Regexp.new(ARGV[0])   # treat command-line arg as regular expression
IO.popen("ps -w -w -x -o'pid,command'") {|ps|
    ps.gets                 # discard header line
    ps.each {|line|
        pid = line.split[0].to_i
        if line =~ pat and pid != Process.pid then
            print line.chomp
            begin
                print "? "
                answer = $stdin.gets
            end until answer =~ /^[yn]/i
            if answer =~ /^y/i then
                Process.kill(9, pid)
                sleep(1)
                begin       # expect exception (process gone)
                    Process.kill(0, pid)
                    $stderr.print("unsuccessful; sorry\n"); exit(1)
                rescue      # handler -- do nothing
                end
            end
        end
    }
}
